### PR TITLE
Update `get_var` to handle missing/None values in a more specific way

### LINKED ISSUE #6277: Update frontend and backend JSON logic implementation
This is to properly check if a value is present in the context, instead of defaulting to `None`/`null`. In the backend, it means putting a custom object as default in the `not_found` argument of `get_var`. Need to implement something similar in the frontend, possibly need a custom override of the "var" operation.

To not break the `!` and `!!` operators, probably need to make sure the custom value is nullish, like the `undefined` value in JS. This also makes sure `{"if": [{"var": "foo"}, ...]}` will still work

### PR DESCRIPTION
**Pull request by vaszig**
Closes https://github.com/open-formulieren/open-forms/issues/6277 partly

### CODE REVIEW

File: json_logic/__init__.py [hunk 1]
```diff
@@ -131,10 +149,8 @@ def get_var(data, var_name, not_found=None):
                 data = data[int(key)]
     except (KeyError, TypeError, ValueError, IndexError):
         return not_found
-    else:
-        if data is None and not_found is not None:
```
Comment (viktorvanwijk) [inline json_logic/__init__.py:135]: I think we need to keep this else statement to ensure we still return the `not_found` value when data is `None`. 
```python
default_not_found = UNDEFINED_VALUE if use_var_undefined else None
if data is None and not_found is not default_not_found
```
Something like this maybe (see also my later comment)

File: json_logic/__init__.py [hunk 2]
```diff
@@ -332,7 +353,27 @@ def jsonLogic(tests: JSON, data: JSON = None, permissive: bool = False) -> JSON:
 
     match operator:
         case "var":
-            return get_var(data, *values)
+            if use_var_undefined:
```
Comment (viktorvanwijk) [inline json_logic/__init__.py]: Because there exists recursion on the above line (and other places), this will only work when the json logic expression is a variable operation. Any nested variable expression (inside an if statement for example) will behave differently. So it needs to be passed to all `jsonLogic` calls

I think the flag `use_var_undefined` is still nice though. Perhaps we can pass this flag to the `get_var` hook to make the distinction which `not_found` value to use there, instead of here?

Still would be good to run the entire shared test suite with `use_var_undefined=True` as well I think. Probably need to skip all the tests that assert the value will be `None` when it's missing

File: json_logic/__init__.py
Comment (viktorvanwijk) [inline json_logic/__init__.py:295]: The undefined value needs to be included here, to prevent performing operations with it (see the last few lines of `jsonLogic`)

File: json_logic/__init__.py [hunk 3]
```diff
@@ -118,10 +136,14 @@ def merge(*args):
...
 
-def get_var(data, var_name, not_found=None):
-    """Gets variable value from data dictionary."""
-    if var_name == "" or var_name is None:
-        return data  # Return the whole data object
+def get_var(data, var_name, not_found=UNDEFINED_VALUE, use_var_undefined=False):
+    if not var_name:
+        return data
+
+    if not use_var_undefined and not_found is UNDEFINED_VALUE:
```
Comment (viktorvanwijk) [inline json_logic/__init__.py:147]: Perhaps with a comment:
Setting `UNDEFINED_VALUE` as the default value for `not_found` function argument ensures we can detect when no value was passed, because it's not possible to specify this undefined value in a JSON logic expression. We need to switch back to `None`, though, if we are opting out of using the undefined value.

File: tests/test_jsonlogic.py [hunk 4]
```diff
@@ -572,3 +572,753 @@ def test_with_permissive(self):
             data = {}
             with self.assertRaises(TypeError):
                 jsonLogic(expression, data, True)
+
+
+class JSONLogicWithVarUndefinedTest(unittest.TestCase):
```
Comment (viktorvanwijk) [inline tests/test_jsonlogic.py:577]: I'm assuming this is a complete copy (with a few additions/changes) of the original test class? I'm wondering if it can be done more efficiently with `hypothesis` or `unittest_parametrize` 🤔 . But we only have those in the backend, so it's OK for now I think

Comment (vaszig) [reply tests/test_jsonlogic.py:577]: Yeah, I checked if we could that with `unittest_parametrize` but thought it's not necessary to add one more dependency for this. Additionally, I think it would not be so friendly for the reader (more "duplicated code" but more clear the reason imo).

**Review by viktorvanwijk — APPROVED** [review]
